be/src/vec/functions/function_string.h

```diff
@@ -553,20 +563,123 @@ class FunctionStringConcatWs : public IFunction {
+            }
+            // Concat string in array
+            _execute_array(input_rows_count, array_column, buffer, views, offsets_list, chars_list,
+                           null_list, res_data, res_offset);
+
+        } else {
+            // Concat string
+            _execute_string(input_rows_count, argument_size, buffer, views, offsets_list,
+                            chars_list, null_list, res_data, res_offset);
+        }
+        if (is_null_type) {
+            block.get_by_position(result).column =
+                    ColumnNullable::create(std::move(res), std::move(null_map));
+        } else {
+            block.get_by_position(result).column = std::move(res);
+        }
+        return Status::OK();
+    }
+
+private:
+    void _execute_array(const size_t& input_rows_count, const ColumnArray& array_column,
+                        fmt::memory_buffer& buffer, std::vector<std::string_view>& views,
+                        const std::vector<const ColumnString::Offsets*>& offsets_list,
+                        const std::vector<const ColumnString::Chars*>& chars_list,
+                        const std::vector<const ColumnUInt8::Container*>& null_list,
+                        Chars& res_data, Offsets& res_offset) {
+        // Get array nested column
+        const UInt8* array_nested_null_map = nullptr;
+        ColumnPtr array_nested_column = nullptr;
+
+        if (is_column_nullable(array_column.get_data())) {
+            const auto& array_nested_null_column =
+                    reinterpret_cast<const ColumnNullable&>(array_column.get_data());
+            // String's null map in array
+            array_nested_null_map =
+                    array_nested_null_column.get_null_map_column().get_data().data();
+            array_nested_column = array_nested_null_column.get_nested_column_ptr();
+        } else {
+            array_nested_column = array_column.get_data_ptr();
+        }
+
+        const auto& string_column = reinterpret_cast<const ColumnString&>(*array_nested_column);
+        const ColumnString::Chars& string_src_chars = string_column.get_chars();
+        const ColumnString::Offsets& src_string_offsets = string_column.get_offsets();
+        const ColumnArray::Offsets& src_array_offsets = array_column.get_offsets();
+        ColumnArray::Offset current_src_array_offset = 0;
+
+        // Concat string in array
         for (size_t i = 0; i < input_rows_count; ++i) {
-            auto& seq_offsets = *offsets_list[0];
-            auto& seq_chars = *chars_list[0];
-            auto& seq_nullmap = *null_list[0];
-            if (seq_nullmap[i]) {
-                res_data.push_back('\0');
+            auto& sep_offsets = *offsets_list[0];
+            auto& sep_chars = *chars_list[0];
+            auto& sep_nullmap = *null_list[0];
+
+            if (sep_nullmap[i]) {
                 res_offset[i] = res_data.size();
```

Comment:
concat_ws(NULL, [1,2,3]) return NULL?

But the impl here returns an empty string.